alternator/streams.cc

```diff
@@ -486,47 +480,130 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
         auto i = topologies.begin();
         auto e = topologies.end();
         auto prev = e;
-        auto threshold = db_clock::now() - confidence_interval(db);
         auto shards = rjson::empty_array();
 
         std::optional<shard_id> last;
 
+        // filter out cdc generations older than the table or now() - 24h
+        auto low_ts = std::max(as_timepoint(schema->id()), db_clock::now() - 24h);
+        while (i != e) {
+            auto j = i;
+            if (++j != e && j->first < low_ts) {
+                i = j;
+                continue;
+            }
+            // next is younger.
+            break;
+        }
+
+        auto first = i;
+
+        if (shard_start) {
```

Comment:
ì`, as conventions go, is the iterator used to extract data. It serves exactly one purpose. It is the generation being processed. We move it forwards to skip irrelevant/already processed generations, but it is still the same purpose iterator.

Reply:
It's a good and obvious convention for maybe a 10-line algorithm where `i` is used as a single loop iterator, but this function has at least a hundred lines.

It's hard to keep track of all these variables. Some of them serve a single purpose but then aren't used for the rest of the function (`low_ts`, `first`), but you don't know that before you read the entire function.

Also, when I read code, I first want to know **what** it does, and only then **how** it does it. Here it's totally not clear what the code is supposed to achieve before you read it entirely. It could be made easier to read if parts of the code had its own names and documentation. For example, before you enter the `for` loop below, you could have something like
`auto [prev, i] = find_first_relevant_generation(topologies)`
where the meaning of "relevant" would be described in a comment (i.e. it may still contain data for this table). And suddenly I only need to read the name of the function and a 2-line comment to know what `prev` and `i` is, I don't even need to read the function's body. Then, when I want to review the function itself, I know what it's supposed to do so it's easier for me to reason about its correctness and easier for me to catch potential bugs.

Reply:
Otoh, creating a subroutine for code not run more than once is rather silly, where a better comment would suffice. I'll opt for the latter.

Reply:
> creating a subroutine for code not run more than once is rather silly

Why? It is known that a human's short-term memory can only keep 5 to 9 items at once. So I'd argue that if a part of the code is easy to be named and separated to a logically-self-contained function (in this case it doesn't modify any state, it just searches for a thing in a set of things), it's good to do it so it won't pollute our short-term memory when we read the rest of the code.